### ISSUE TITLE
Very suspicious condition evaluation and error 🪄

### ISSUE THREAD

**Issue by ExtReMLapin** [ArcadeData/arcadedb#1563]
Hello,

(TLDR at the bottom)

Following issue #1562 , same database, same environement.

The database is available in #1562 

As a reminder here was the query ran by my coworker.

```
MATCH (n) WHERE ('NER' IN labels(n) OR 'THEME' IN labels(n)) AND (toLower(n.identity) CONTAINS 'technology' AND toLower(n.identity) CONTAINS 'science' AND toLower(n.identity) CONTAINS 'notable advancements' AND toLower(n.identity) CONTAINS '2020') RETURN n.identity as identity, ID(n) as id
```

This query can be simplified and fixed a little bit by using ` (n:NER OR n:THEME)` instead of using `labels()`

Now this specific query errors: 

```CYPHER
MATCH (n) WHERE (n:NER OR n:THEME) AND
(
    toLower(n.identity) CONTAINS 'technology' AND
    toLower(n.identity) CONTAINS 'science' AND
    toLower(n.identity) CONTAINS 'notable advancements' AND
    toLower(n.identity) CONTAINS '2020'
)
RETURN n
```

What's really weird is that all NER and all THEME nodes should contain identity property, so it should not error, so maybe the issue reported in #1562 is another bug that the one we're metting here, as the two following querries works fine 

```CYPHER
MATCH (n:NER) WHERE
(
    toLower(n.identity) CONTAINS 'technology' AND
    toLower(n.identity) CONTAINS 'science' AND
    toLower(n.identity) CONTAINS 'notable advancements' AND
    toLower(n.identity) CONTAINS '2020'
)
RETURN n
```
(returns nothing but no error)

```CYPHER
MATCH (n:THEME) WHERE
(
    toLower(n.identity) CONTAINS 'technology' AND
    toLower(n.identity) CONTAINS 'science' AND
    toLower(n.identity) CONTAINS 'notable advancements' AND
    toLower(n.identity) CONTAINS '2020'
)
RETURN n
```
(returns nothing but no error)

So the first conclusions we can have here is that individually running this query on NER and THEME nodes classes works but both at the same time isn't working ?

My conclusion here was that the Cypher engine is evaluating all the conditions even if the first one (the one that check if it's  `NER` or a `THEME`) doesn't "pass".

Why ? because when I run this query : 

```CYPHER
MATCH (n) WHERE n.identity IS NULL return n
```

returns only `IMAGE` nodes where there is no `identity` field/property.

So from there, my conclusions was as stated before that it was evaluations all conditions ignoring the order.

# TLDR START

Howhever, this is where is gets weird.

Remember this query ? 

```CYPHER
MATCH (n) WHERE (n:NER OR n:THEME) AND
(
    toLower(n.identity) CONTAINS 'technology' AND
    toLower(n.identity) CONTAINS 'science' AND
    toLower(n.identity) CONTAINS 'notable advancements' AND
    toLower(n.identity) CONTAINS '2020'
)
RETURN n
```

When, when you remove ANY of the 4 conditions between the parentheses, it doesn't error anymore.

**Comment by ExtReMLapin** [comment]
https://github.com/ArcadeData/arcadedb/assets/3909752/28dac7b7-127a-403c-9092-a5909b71ffcf

**Comment by ExtReMLapin** [comment]
It really doesn't make any sense.

Execute the query 4 times, will error on the 4th time, not the 3 first times.

New query : 

```CYPHER
MATCH (n)
WHERE (n:NER OR n:THEME)
  AND (n.identity IS NOT NULL)
  AND (toLower(n.identity) CONTAINS 'science')
  AND (toLower(n.identity) CONTAINS 'notable advancements')
  AND (toLower(n.identity) CONTAINS '2020')
  AND (toLower(n.identity) CONTAINS '2021')
RETURN n
```
Video here where i'm just sending the same request OVER and OVER, sometime it fails, sometimes it doesn't.

https://github.com/ArcadeData/arcadedb/assets/3909752/a2123709-7298-486e-a7d0-ddd3c8c3a331

**Comment by lvca** [comment]
I tried al your queries against the provided database but I don't see any error against the latest release. Have you tried it with the latest `main` branch with yesterday's fix? I'm assuming the same bug was responsible for this.

**Comment by ExtReMLapin** [comment]
Thanks for the quick answer, I downloaded the latest snapshot `arcadedb-package-24.5.1-20240423.041740-4.zip` , didn't know there were automatic builds.

However the snapshot doesn't include cypher and studio, and merging with latest release doesn't help as it reports the release version, not the headless one, do you have a 90mb full release somewhere ?

**Comment by ExtReMLapin** [comment]
As for yesterday's fix, I can confirm the issue is not happening anymore.
As for this issue and the very weird behavior cited above, I do not meet this issue anymore with the snapshots, weird.

I feel there was still an underlying bug, if the **first** condition was correctly evaluated (only filter `NER` and `THEME`) how could it error then ? 

Don't you think there is something else going on ? The error in the first place should never have happened as all NER and THEME nodes have the "indentity" property.

**Comment by ExtReMLapin** [comment]
>About the later question, the cypher module couldn't be that "smart" and execute all the conditions. Or maybe the conditions are not kept in a list but shuffled (in a set)? So the order can be pseudo-casual?

The later theory sound plausible as as I showed in the video, running the same query 10 times will **not** cause an error 10 times.
That sounds really problematic as the [Short-circuit evaluation](https://en.wikipedia.org/wiki/Short-circuit_evaluation) principle/rule doesn't seem to be respected. 

Again, thank you for the time spent improving ArcadeDB

**Comment by lvca** [comment]
Unfortunately, the Cypher module is a separate module, and the OpenCypher team seems not interested in working on it. ALl we can do is little patches until somebody comes up with a complete rewrite.
